### PR TITLE
IE Non-Enumerable properties

### PR DESCRIPTION
**Pull request by jridgewell**
I’m not sure what we were doing before with the `FuncProto`…

@jdalton: mind giving this a review?

### CODE REVIEW

**Comment by paulfalgout** [comment]
This seems to work in IE8.  Beat me to it.  If you're waiting for @jdalton to review.. it might be awhile..  he's been banned.

**Comment by jridgewell** [comment]
I've got it working in IE6 and 8, so I think we're set.

> If you're waiting for @jdalton to review.. it might be awhile.. he's been banned.

Well, it's been one of those days.

**Comment by jashkenas** [comment]
What's the difference from the previous implementation?

**Comment by jridgewell** [comment]
The only important change is determining the `prototype` to compare to.

```diff
@@ -894,17 +896,21 @@

   // Keys in IE < 9 that won't be iterated by `for key in ...` and thus missed.
   var hasEnumBug = !{toString: null}.propertyIsEnumerable('toString');
-  var nonEnumerableProps = ['constructor', 'valueOf', 'isPrototypeOf', 'toString',
+  var nonEnumerableProps = ['valueOf', 'isPrototypeOf', 'toString',
                       'propertyIsEnumerable', 'hasOwnProperty', 'toLocaleString'];

   function collectNonEnumProps(obj, keys) {
     var nonEnumIdx = nonEnumerableProps.length;
-    var proto = typeof obj.constructor === 'function' ? FuncProto : ObjProto;
+    var constructor = obj.constructor;
+    var proto = (_.isFunction(constructor) && constructor.prototype) || ObjProto;
+
+    // Constructor is a special case.
+    var prop = 'constructor';
+    if (_.has(obj, prop) && !_.contains(keys, prop)) keys.push(prop);

     while (nonEnumIdx--) {
-      var prop = nonEnumerableProps[nonEnumIdx];
-      if (prop === 'constructor' ? _.has(obj, prop) : prop in obj &&
-        obj[prop] !== proto[prop] && !_.contains(keys, prop)) {
+      prop = nonEnumerableProps[nonEnumIdx];
+      if (prop in obj && obj[prop] !== proto[prop] && !_.contains(keys, prop)) {
         keys.push(prop);
       }
     }
```

**Comment by paulfalgout** [comment]
The only functional change I see really is using `obj.constructor.prototype` vs `FuncProto`

**Comment by jashkenas** [comment]
Kill the "fasley"s?

**Comment by megawac** [comment]
LGTM :+1:

**Comment by jridgewell** [comment]
> The only functional change I see really is using obj.constructor.prototype vs FuncProto

That's what I meant.

> Kill the "fasley"s?

But they pass? And it's good to test edge cases.

- - -

Travis is erroring because of issues with https://github.com/jashkenas/underscore/commit/173a7fa817e17b3cc5d555180ba0ea4d35655172. https://github.com/jashkenas/underscore/pull/2065

**Comment by paulfalgout** [comment]
> That's what I meant.

You're too quick for me

The tests should say "falsey" I think right?

File: test/objects.js [hunk 1]
```diff
@@ -736,6 +772,10 @@
 
     Prototest.x = 5;
     ok(_.isMatch({x: 5, y: 1}, Prototest), 'spec can be a function');
+
+    //null edge cases
+    var oCon = {'constructor': Object};
+    deepEqual(_.map([null, undefined, 5, {}], _.partial(_.isMatch, _, oCon)), [false, false, false, true], 'doesnt fasley match constructor on undefined/null');
```
Comment (JakeChampion) [inline test/objects.js]: `"fasley"` should be `"falsey"`

Comment (jridgewell) [inline test/objects.js]: Already [fixed](https://github.com/jashkenas/underscore/pull/2066).
